New terms_enum API for discovering terms in the index.

markharwood opened the pull request:
A search string is supplied which is used as prefix for matching terms found in a given field in the index.
A timeout can limit the amount of time spent looking for matches.
Designed for use in Kibana auto-complete use cases.
Kibana requests for this API would typically look like this:
```
localhost:9200/myindex/_terms_enum
{
"field" : "myfield",
"string" : "Microsof",
"index_filter": {
    "bool":{
        "must":[
           { "range": "my Kibana time-picker range"},
           { "terms": {"_tier": ["data_warm", "data_hot"]}
         ]
}}
```
The time range would avoid any indices that fall outside of the range _but does not filter any doc values in overlapping indices_. The tier clause would avoid hitting frozen/cold indices.
An optional `timeout` [time value](https://www.elastic.co/guide/en/elasticsearch/reference/7.12/common-options.html#time-units) can also be passed  (default is "1s", one second). 

The response looks like this:
```
{
  "_shards": {
    "total": 1,
    "successful": 1,
    "failed": 0
  },
  "terms": [
    "Microsoft",
    "Microsoft Windows"
  ],
  "complete": true
}
```
Any requests that hit the `timeout` setting will return `"complete":false`

markharwood committed:
A TermsEnum API for discovering terms in the index.
Terms matching a given prefix can be returned in alphabetical or by popularity,
A timeout can limit the amount of time spent looking for matches.

Expected to be useful in auto-complete or regex debugging use cases.

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

mayya-sharipova commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TermCount.java:
```diff
@@ -0,0 +1,107 @@
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        TermCount other = (TermCount) o;
+        return Objects.equals(getTerm(), other.getTerm()) && Objects.equals(getDocCount(), other.getDocCount());
```
since `getDocCount()` returns a primitive type, may be for comparison we can just do: `getDocCount() == other.gertDocCount()`?

mayya-sharipova commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TermEnumResponse.java:
```diff
@@ -0,0 +1,119 @@
+
+    public static final String TERMS_FIELD = "terms";
+    public static final String TIMED_OUT_FIELD = "timed_out";
+    public static final String SKIPPED_FIELD = "skipped";
```
what does `SKIPPED_FIELD` is responsible for ?

markharwood replied:
Like the search API - it's what shards failed the canMatch phase.
If you supply an `index_filter` it does a canMatch on each shard before considering the terms in it. (Currently the test fails for use of index filters after I moved the code to xpack. Something about QueryBuilders not being available).

markharwood replied:
Removed

mayya-sharipova commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TermEnumAction.java:
```diff
@@ -0,0 +1,51 @@
+
+    public static final TermEnumAction INSTANCE = new TermEnumAction();
+//    public static final String NAME = "indices:admin/termsenum/list";
+    public static final String NAME = "indices:data/read/xpack/termsenum/list";
```
May be too early to talk about this, but when security is enabled, who can access to this endpoint (those who can read the index))?  what about when FLS and DLS is enabled?

markharwood replied:
DLS and FLS is not part of this PR currently. A TBD.

mayya-sharipova replied:
@markharwood Thanks for the answers, Mark. Is this PR is ready for a more thorough review? or you wanted to add more/ do more modifications?

markharwood replied:
Thanks for the review Mayya! There's still some stuff I want to add just yet. (Sorry, I should have added the WIP label).
I need to add logic and tests to filter slower indices based on tiers (we ignore frozen indices and searchable snapshots).

I'll leave DLS and FLS for a subsequent PR.

jimczi replied:
Should it be `indices:data/read/terms` ? No need to add xpack reference here.

mayya-sharipova reviewed (commented):
@markharwood Exciting API ,  I envision it to be very useful .  I did an initial pass in review, mostly questions.

markharwood commented on docs/reference/search/term-enum.asciidoc:
```diff
@@ -0,0 +1,130 @@
+
+[source,console]
+--------------------------------------------------
+POST stackoverflow/_terms/list
```
No plans but I thought it might make sense to leave things open to that

markharwood commented:
When it comes to the HLRC I'm unsure where best to place the logic:

1) Do I make a copy of Request/Response classes for use in HLRC? (Helps long term goal of decoupling server classes from client).
2) Do I create a new XxxxxClient class or add methods to existing class e.g. IndicesClient? (Bear in mind this new method is basic-licensed and perhaps shouldn't be mixed in with OSS methods).

jimczi reviewed (changes requested):
I did a first pass to review the main API and the options.
I think we're mixing multiple use cases and allow too many options. I left some comments to simplify the API.  I'll make a second one to review the concrete actions in a follow up.

markharwood commented on docs/reference/search/term-enum.asciidoc:
```diff
@@ -0,0 +1,130 @@
+`_all` or `*`.
+
+[[search-term-enum-api-request-body]]
+==== {api-request-body-title}
```
The design doc from @giladgal  had a stated objective of infix searching (perhaps for a later phase): 

>"The use case to keep in mind for this scenario would be URLs and a user that types the domain name without the ‘http://www’ prefix or the .co suffix (perhaps not knowing the exact prefix and suffix)."

I started with the search string called `prefix` but changed it to `pattern` to accommodate these plans for other search modes.

jimczi commented on docs/reference/search/term-enum.asciidoc:
```diff
@@ -0,0 +1,101 @@
+Wildcard (`*`) expressions are supported.
++
+To search all data streams or indices in a cluster, omit this parameter or use
+`_all` or `*`.
```
Should we forbid searching all data ? Plain `*` or omitting the pattern could return an error ?

markharwood replied:
We have the tier filtering now so this so searching all may be more viable

jimczi commented on docs/reference/search/term-enum.asciidoc:
```diff
@@ -0,0 +1,101 @@
+  "terms": [
+    "kibana"
+  ],
+  "timed_out" : false
```
We also need to know if the result is complete or not ? For small cardinality field it can be useful to know that the response contains all the possible values for that input.

markharwood replied:
"complete = true" will have some caveats e.g. if we skip indices on cold nodes

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/ShardTermEnumRequest.java:
```diff
@@ -0,0 +1,129 @@
+        return field;
+    }
+
+    public String pattern() {
```
You need to change the name here too (`string`)  ?

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java:
```diff
@@ -0,0 +1,595 @@
+        this.transportService = transportService;
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+        this.transportShardAction = actionName + "[s]";
+        // TODO new threadpool
```
Is the TODO still relevant ? The threadpool is there from what I can see.

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java:
```diff
@@ -0,0 +1,595 @@
+    protected GroupShardsIterator shards(ClusterState clusterState, TermEnumRequest request, String[] concreteIndices) {
+        final String routing = null;
+
+        // Remove any cold or frozen indices from the set of indices to be searched.
```
We should rely on node's data-tier when selecting indices and not care about index settings. The goal is to   restrict the API to some data-tiers, we can add more flexibility in the future but I'd prefer that we start simple. Only the hot and warm tier should be queried.

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java:
```diff
@@ -0,0 +1,595 @@
+                }
+            }
+
+        } catch (AssertionError e) {
```
we should never catch an AssertionError, did you mean `Exception` ?

markharwood replied:
Yep - a copy and paste from TransportValidateQueryAction

jimczi commented on server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -171,6 +172,7 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
         builders.put(Names.GET, new FixedExecutorBuilder(settings, Names.GET, allocatedProcessors, 1000, false));
         builders.put(Names.ANALYZE, new FixedExecutorBuilder(settings, Names.ANALYZE, 1, 16, false));
         builders.put(Names.SEARCH, new FixedExecutorBuilder(settings, Names.SEARCH, searchThreadPoolSize(allocatedProcessors), 1000, true));
+        builders.put(Names.AUTO_COMPLETE, new FixedExecutorBuilder(settings, Names.AUTO_COMPLETE, 2, 10, true));
```
In the interest of limiting the number of threads, I wonder if we can group the shard requests per node ?
If there are 30 shards to lookup on a single node, I think it would be better to do it in a single call and adapt the strategy globally. With the current configuration, running the request on these 30 shards would raise 18 errors out of 30 (2 can run and 10 are queued, the rest are refused).

markharwood replied:
Would it best to leave unbundled, then admins can play with degrees of parallelism they want to throw at the problem by tweaking num threads and queue size settings?

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java:
```diff
@@ -0,0 +1,595 @@
+            request.taskStartedTimeMillis(),
+            request.filteringAliases()
+        );
+        SearchContext searchContext = searchService.createSearchContext(shardSearchLocalRequest, SearchService.NO_TIMEOUT);
```
You don't need a `SearchContex` here, you should use a`QueryShardContext` instead like we do in the can match phase. Also note that you could use the same `QueryShardContext` for the canMatchShard above, `SearchService#queryStillMatchesAfterRewrite` was added for this purpose.

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java:
```diff
@@ -0,0 +1,595 @@
+                    if (node == null) {
+                        // no node connected, act as failure
+                        onOperation(shard, shardIt, shardIndex, new NoShardAvailableActionException(shardIt.shardId()));
+                    } else {
```
The timeout could be checked here to save some round trip and could avoid accumulating requests on the node's thread pool ?

jimczi reviewed (commented):
Sorry for the delay.
I left another round of comments regarding the _terms  action.

mayya-sharipova commented on docs/reference/search/term-enum.asciidoc:
```diff
@@ -0,0 +1,101 @@
+Which field to match
+
+[[term-enum-string-param]]
+`field`::
```
should this be `string`?

markharwood replied:
Good catch!

markharwood commented:
@elastic-jb Just updated this PR if you want to AB test the sort-by-popularity with current Kibana terms-agg approach.

This PR only considers up to a max of 10k matching terms on a node and returns `complete : false` if that limit is reached.
This is not a guaranteed API at this stage so is being made available for evaluating results quality/speed with the existing terms agg approach.
If you want to compare results interactively - I have a Python Flask webserver you can run. It lets you pick an index+field:
<img width="156" alt="Select_index" src="https://user-images.githubusercontent.com/170925/106741492-ab57c580-6613-11eb-9e9b-5127c0ad1929.png">
then try the various searches like this search for bands starting with `A`:

<img width="291" alt="Home_-_Microblog" src="https://user-images.githubusercontent.com/170925/106741895-3a64dd80-6614-11eb-9790-6667055bc051.png">

Note, in the above example of music artists sorted by popularity, the results are incomplete because there are more than 10k bands starting with `a`.  Typing any additional character reduces the matches and makes the scan return `complete:true`.
Even when incomplete I expect the accuracy of these results will be greater than the current terms-agg approach (which is based on a sample of the first random N docs). I expect the speed to be significantly faster with this PR too.
I'll do some benchmarks on various datasets to measure the speed and quality difference between this PR and existing terms-agg approach. Comparing speed is easy. The quality is harder but should be measurable by the sorts of measures used in traditional search results evaluation - the deviation of a candidate configuration from a "gold standard" set of ordered results. In this case the gold standard is not some human-judged ranking of quality but something we can produce automatically - a strict doc_count popularity order produced by fully accurate terms aggs that are not subject to any time or space limits.

markharwood commented:
@jimczi @elastic-jb @lizozom @giladgal 

I tried this PR out compared with the existing Kibana impl (terms agg with `terminate_after:100k docs`).
To compare like-with-like I used a sort-by-popularity mode but with a 10k limit on number of matching terms considered in the index.

I had plans to automate the comparisons but there were some interesting findings just from manually playing with some datasets.

## 1) The existing Kibana impl is case sensitive
Unless the user has chosen to add normalisation to the mapping at index-time, there is no case insensitivity to searches: 
<img width="312" alt="Dallas crimes - Kibana case sensitive" src="https://user-images.githubusercontent.com/170925/106784624-70ba5100-6644-11eb-9048-7a8a8b089c50.png">
Note the existing Kibana impl could be changed to make the `include` regex it uses to filter terms case insensitive but, as we'll see, there are other concerns.

## 2) The existing Kibana impl misses stuff
Sampling the first 100k docs has no guarantee of finding the term the user is looking for as in this example:
<img width="311" alt="Dallas crimes - Kibana missing data" src="https://user-images.githubusercontent.com/170925/106785040-e6262180-6644-11eb-9fce-e7780340bf83.png">

## 3) The existing Kibana impl is slow in ways we didn't realise
One of my datasets has many values per doc - person profiles have a list of band "likes" related to the user. This acts as big multiplier to the number of regex tests the `terms` agg does on the 100k docs in the sample. This dataset was noticeably slower than my tests on other datasets with similar numbers of docs but which had single-value fields. . It doesn't matter how long the searched prefix is - this took nearly 2 seconds for every request so is a fixed cost. The new impl was much faster:
<img width="466" alt="Music fans_-_Kibana multivalue fields slow" src="https://user-images.githubusercontent.com/170925/106785540-651b5a00-6645-11eb-8c14-d1ed083773ff.png">

## 4) The new impl can be inaccurate (but there's a simple solution)
For a high cardinality field and a short search string the results from the new impl can be inaccurate. In this example search for `a` there are many bands with that prefix so rather than considering `aa` to `az` we hit the fixed 10k limit of terms somewhere around `am`. We still found popular bands like `abba` but are missing the most  popular `arctic monkeys`. This incompleteness is flagged in the response:
<img width="333" alt="Music fans - new api misses top" src="https://user-images.githubusercontent.com/170925/106786077-0e625000-6646-11eb-862a-b00bb942654b.png">
However, the solution is simple - when the result is incomplete the user should be encouraged to "type more" and, in this case,  with a single key press the results become fully accurate and are faster:

<img width="301" alt="Music fans - new api longer strings are same" src="https://user-images.githubusercontent.com/170925/106786338-584b3600-6646-11eb-9948-68eb1ab28683.png">
This is a big improvement on the existing implementation where typing more characters neither improves speed or accuracy.

jtibshirani commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java:
```diff
@@ -0,0 +1,605 @@
+            }
+            MultiReader multiReader = new MultiReader(shardReaders.toArray(new IndexReader[0]), false);
+
+            Terms terms = MultiTerms.getTerms(multiReader, request.field());
```
I realize this is WIP, but had one high-level thought that might affect code structure/ naming. It'd be great to encapsulate this term loading logic into `MappedFieldType` instead of interacting with Lucene data structures directly here. We have been trying to do this consistently, and now delegate query creation, doc value loading, and other data fetching tasks to `MappedFieldType`. Some benefits:
* Field types can decide to handle the call differently. For example `constant_keyword` could support this functionality (even though it doesn't write them in the index).
* Field aliases will be handled correctly, since we resolve them when looking up `MappedFieldType`
* If there's any change to logic because of an index version change, we can consolidate that in one place

markharwood replied:
I did suggest that MappedFieldType delegation to Jim the other day but he was not so keen but I don't recall exactly why.
Possibly deferring the change to a later development phase or a concern over losing the optimisation we get from having a single MultiReader view on multiple shards/indices. Do you remember @jimczi ?

jimczi replied:
+1 to add the logic to return `Terms` or `TermsEnum` in the `MappedFiedlType`. Mark we've discussed this already and I remember saying that it was important in order to handle `constant_keyword` and `flattened` correctly ;).
It could return an empty or null terms by default to make it clear that the API is not mandatory for each field type.

markharwood replied:
What's your feeling on the timer question in my (edited) comment above?

jimczi replied:
I don't understand the relation with completeness ? We're talking of a terms enum that can only be used to search by prefix so the proposal is something like:
`TermsEnum getTerms(String prefix, boolean caseSensitive)` or `Terms getTerms()` where we'd leave more room for the caller. So there's no need for each field type to implement anything regarding early termination or timeout. It's the terms enum API that wraps all these `TermsEnum` in a `MultiTermsEnum` and knows where to stop and if the result is complete or not.

markharwood replied:
It's allowing for anything that might be expensive e.g. runtime fields or wildcard fields that choose to attempt _something_ within the provided time constraints (timeout being something you said should be a parameter in this API).
They'd have to gather a sample of docs _before_ returning a TermsEnum which enumerates over the discovered terms.

romseygeek replied:
I think runtime fields should just return an empty TermsEnum here? Probably the same with wildcard fields tbh, they don't have a sensible terms dictionary to respond from.  Or wildcard fields could return a MultiTerms over the term dictionaries for all their doc value leaf dictionaries, which will be fast to construct even if its slower to iterate over

markharwood replied:
I don't think wildcard could do anything _trustworthy_ whether it chooses to block on the construction or the logic that determines what to return first from terms.next();
That's kind of the point here - we want to communicate to users if there's the possibility of any false negatives in the results being presented. A false negative is bad because we don't want users to walk away without searching because the autocomplete results have (falsely) suggested to them there's nothing to be found.  
We have several policy options:
1) Kibana teaches users to distrust _all_ suggestions
2) Users generally trust results but Kibana flags any incomplete suggestions. Field types inform incompleteness by either: 
  a) returning all-or-nothing. A null response means users are told there's stuff missing. A non-null response means the set is assumed to be complete
  b) fields types return a subset of  terms and somehow communicates the completeness of this set.

If we assume completeness is useful (and that was a design goal) we need to pick either 2a or 2b for this API contract

jimczi replied:
> It's allowing for anything that might be expensive e.g. runtime fields or wildcard fields that choose to attempt something within the provided time constraints (timeout being something you said should be a parameter in this API).
They'd have to gather a sample of docs before returning a TermsEnum which enumerates over the discovered terms.

Please no, we discussed that already and we said that it's not something we want to provide. If you have a wildcard or a runtime field then you're not eligible to the terms enum API and we return fields. That's the part that should be clear in the javadocs of the method, return the terms enum only if you can provide the feature efficiently.

markharwood replied:
OK. What confused me was the idea that timeout should be a parameter passed to this field type API and how that would be interpreted/used

markharwood replied:
The MappedFieldType interface may need some more thought.
For efficiency's sake the current PR uses a MultiReader on a node to provide a single view across potentially many shards. While quick, there are several issues here:

1) **Integer doc count limit** - if we ever want to sort by popularity, the TermsEnum.docFreq() returns an integer which will potentially overflow if we try use a single view across multiple shards. Using this Lucene-based abstraction over multiple Lucene indices is not going to scale (TODO - fix in Lucene or copy Lucene's MultiTermsEnum ?)
2) **Resource management** the current one-size-fits-all logic  has a try-finally block to close readers used while iterating across Terms - the proposed new MappedFieldType API above has not considered how to close any underlying resources accessed by the returned Terms object.
3) **Some, all or no fusion delegation?** - an open question is if the underlying field types are expected to perform any fusion of data across shards in the view offered by the returned TermsEnum. How do they know which Lucene  index or indices they should access? Are they passed one shardID or multiple? How would the caller determine that 2 field types e.g. the `keyword` field for `host` in index `logs08-02-2021` and the same field in index `logs09-02-2021` were merge-able by one of the field types? We would, of course, assume that MappedFieldTypes of different classes would have different implementations and so their TermsEnums would have to be merged by the caller doing the collection on a node. How would the caller fuse multiple TermEnums? The Lucene MultiTermsEnum class is constructed using `ReaderSlice` objects.

markharwood replied:
Discussed with Jim. We'll go with

    TermsEnum getTerms(String prefix, boolean caseSensitive, SearchExecutionContext context)

The caller calls this on the relevant field type for every shard and merges the TermsEnum objects using a new MultiTermsEnum class that differs from the Lucene one in that it:
1) Takes a number of TermsEnum objects rather than ReaderSlice objects
2) Uses longs not ints for summing doc frequencies
3) Doesn't implement TermsEnum, so doesn't offer all its methods like ord, postings etc Instead it has next(), term() and docFreq() only

markharwood committed:
Added HLRC support and related integration test

markharwood committed:
Added client classes for HLRC.
Added yaml test with index filter

markharwood committed:
License fix

markharwood commented:
Two things came out of discussions today

1) We should permanently remove the logic in this PR about filtering tiers once we have the canMatch test supporting[ a new query for testing the tier an index is on](https://github.com/elastic/elasticsearch/issues/68135)
2) We should remove the HLRC support until we can move the code out of xpack (currently we call security APIs directly to avoid DLS indices but to avoid this we'll need a new security API to filter indices where DLS is on for the current user)

markharwood committed:
Remove HLRC code for now - requires less-than-ideal package names while we’re unable to move main server implementation out of xpack (due to security dependency).
Don’t want clients to build dependencies on the wrong package names and the Java client changes are underway anyway.

markharwood committed:
Return empty arrays when no results rather than no `terms` property at all

markharwood commented:
I opened https://github.com/elastic/elasticsearch/issues/70221 to try clean up the custom security logic required in this PR

markharwood commented:
>We should permanently remove the logic in this PR about filtering tiers once we have the canMatch test supporting a new query for testing the tier an index is on

One thing about the above design is that cold nodes will be needlessly bombarded with "can match" requests triggered by Kibana user keystrokes. It's only on the data node that we discover an index is inappropriate. The current implementation filters inappropriate indices at the coordinating node (using hardcoded logic but could inspect a `routing` request param instead of delegating a can-match query). Thoughts, @jimczi ?

jimczi commented:
> The current implementation filters inappropriate indices at the coordinating node (using hardcoded logic but could inspect a routing request param instead of delegating a can-match query). Thoughts, @jimczi ?

We group shards by node so I don't think it's an issue ? I mean we're not "bombarding" nodes with lots of request, just one per node that can take a very quick decision and return immediately if the filter is rewritten to match_none.

markharwood commented:
>I mean we're not "bombarding" nodes with lots of request, just one per node that can take a very quick decision and return immediately if the filter is rewritten to match_none.

It seems a bit wasteful because we never learn  - we just keep going back asking the same old questions - are you warm? No? OK. Be back to ask again soon....

markharwood committed:
Fix bundling of shardIds for nodes, add success/fail accounting of numbers of shards

markharwood committed:
Type fixes

markharwood committed:
Types warning

markharwood committed:
Removed hot/warm tier tests (in anticipation of new queryable _tier field)
Move canMatch logic to run on network thread
Injected searchService so we can use its canMatch method

markharwood commented:
@lizozom Kibana requests for this API will look like this:
```
localhost:9200/myindex/_terms
{
"field" : "myfield",
"string" : "Microsof",
"index_filter": {
    "bool":{
        "must":[
           { "range": "my Kibana time-picker range"},
           { "terms": {"_tier": ["data_warm", "data_hot"]}
         ]
}}
```
The time range may be **optional** - it just avoids any indices that fall outside of the range _but does not filter any doc values in overlapping indices_. The tier clause is **required** to avoid hitting frozen/cold indices.
An optional `timeout` value in milliseconds can also be passed  (default is 1000). 

The response looks like this:
```
{
  "_shards": {
    "total": 1,
    "successful": 1,
    "failed": 0
  },
  "terms": [
    "Microsoft",
    "Microsoft Windows"
  ],
  "complete": true
}
```
Any requests that hit the `timeout` setting will return `"complete":false`

markharwood committed:
Move rest-api-spec and related YML test to new standard home for this stuff.

markharwood committed:
Unused import

markharwood committed:
Move test to xpack

markharwood committed:
Return early on network thread if can’t match any shards.
Remove FLS logic now we use the right searcher.

markharwood committed:
Removed sort by popularity option

markharwood committed:
Unused import

markharwood commented:
Ready for another look when you have the time @jimczi

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java:
```diff
@@ -0,0 +1,552 @@
+        return new NodeTermEnumResponse(request.nodeId(), termsList, error, true);
+    }
+
+    // TODO remove this so we can shift code to server module - see https://github.com/elastic/elasticsearch/issues/70221
```
Is this still accurate. I thought that we would try to rewrite the role query and if it results in a match_all then we'd accept the request ?

markharwood replied:
The TODO should perhaps be about moving any security filtering logic like this to a custom "Interceptor" class which wraps the core implementation and rewrites requests rather than having security+impl in the same class.

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java:
```diff
@@ -0,0 +1,552 @@
+        }
+        if (request.shardIds().size() == 0) {
+            listener.onResponse(new NodeTermEnumResponse(request.nodeId(), Collections.emptyList(), null, true));
+        } else {
```
We could also use the search thread pool if the queue is empty. Happy to do it in a follow up.

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/rest/RestTermEnumAction.java:
```diff
@@ -0,0 +1,53 @@
+            return channel ->
+            client.execute(TermEnumAction.INSTANCE, termEnumRequest, new RestToXContentListener<>(channel));
+        }        
+        
```
nit: extra lines

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TermEnumAction.java:
```diff
@@ -0,0 +1,45 @@
+
+import static org.elasticsearch.index.query.AbstractQueryBuilder.parseInnerQueryBuilder;
+
+public class TermEnumAction extends ActionType<TermEnumResponse> {
```
Should we use the plural (Terms) instead ? That would be consistent with the name of the API (`_terms`). Same for all the other class names (`TermEnumResponse`, ...).

markharwood replied:
I don't know if "Terms" will be too short. We already have a lot of `o.e.Term*` classes to navigate and users looking for help with the `terms` aggregation. 
A verb would be a good addition e.g. termFinder

markharwood replied:
No have REST endpoint `_terms_enum` and `TermsEnumXxxx` class names

jimczi reviewed (commented):
The API looks great @markharwood ! I left some comments and ideas for follow ups. 
@jpountz can you take a look at the high level API when you have time ? We tried to reduce the number of options to a minimum so would be great to get your thoughts before we merge.

markharwood commented on server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:
```diff
@@ -399,4 +400,20 @@ public TextSearchInfo getTextSearchInfo() {
+     * This method is used to support auto-complete services and implementations
+     * are expected to find terms beginning with the provided string very quickly.
+     * If fields cannot look up matching terms quickly they should return null.  
+     * The returned TermEnum should implement next(), term() and doc_freq() methods
```
I kept it in for now because 1) it's cheap to get from the index and 2) Even if we don't use it for popularity sorting we might want to use it later to filter low-frequency items e.g. user tags that occur only once

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/NodeTermEnumRequest.java:
```diff
@@ -0,0 +1,160 @@
+    public NodeTermEnumRequest(StreamInput in) throws IOException {
+        super(in);
+        // Set the clock running as soon as we appear on a node.
+        nodeStartedTimeMillis = System.currentTimeMillis();
```
it feels inconsistent to set it when deserializing, could we require callers to call `shardStartedTimeMillis` instead?

markharwood replied:
Can we rely on nodes' clocks not drifting by the small amounts we measure here?

jimczi replied:
I think what Adrien meant is that we shouldn't set the timer when deserializing the data. That feels fragile and inconsistent. If that represents the time when the action started on a data node then we should be able to set it once in `TransportTermsEnumAction#asyncNodeOperation` ?

markharwood replied:
My goal was to start the timer at the earliest opportunity on the data node (ideally prior to any queuing etc) and given the NodeTermEnumRequest was passed to `TransportTermsEnumAction#asyncNodeOperation` then the deserialization would have occurred before this point?

jimczi replied:
I don't think this is needed. We call `asyncNodeOperation` early enough to not worry too much about what happens before.

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/NodeTermEnumRequest.java:
```diff
@@ -0,0 +1,160 @@
+            nodeStartedTimeMillis = System.currentTimeMillis();
+        }
+        return this.nodeStartedTimeMillis;
+    }    
```
I'm confused why the method is called `shardStartedXXX` while the field is called `nodeStartedXXX`.

markharwood replied:
A relic from when we did per-shard requests rather than bundling into a per-node request

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/NodeTermEnumRequest.java:
```diff
@@ -0,0 +1,160 @@
+        out.writeVInt(size);
+        // Adjust the amount of permitted time the shard has remaining to gather terms. 
+        long timeSpentSoFarInCoordinatingNode = System.currentTimeMillis() - taskStartedTimeMillis;
+        assert timeSpentSoFarInCoordinatingNode >= 0;
```
I don't think that this assert is legal. The delta could be negative if we spent a long time on the above line for instance.

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/NodeTermEnumRequest.java:
```diff
@@ -0,0 +1,160 @@
+        // Adjust the amount of permitted time the shard has remaining to gather terms. 
+        long timeSpentSoFarInCoordinatingNode = System.currentTimeMillis() - taskStartedTimeMillis;
+        assert timeSpentSoFarInCoordinatingNode >= 0;
+        int remainingTimeForShardToUse = (int) (timeout - timeSpentSoFarInCoordinatingNode);
```
why do we need to cast to an int?

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/SimpleTermCountEnum.java:
```diff
@@ -0,0 +1,121 @@
+ * {@link MappedFieldType#getTerms(boolean, String, org.elasticsearch.index.query.SearchExecutionContext)}
+ * but can't return a raw Lucene TermsEnum.
+ */
+public class SimpleTermCountEnum extends TermsEnum{
```
```suggestion
public class SimpleTermCountEnum extends TermsEnum {
```

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/SimpleTermCountEnum.java:
```diff
@@ -0,0 +1,121 @@
+    
+    public SimpleTermCountEnum(TermCount[] terms) {
+        sortedTerms = terms;
+        Arrays.sort(sortedTerms, Comparator.comparing(TermCount::getTerm));
```
could you copy the input array instead of modifying it in-place?

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/SimpleTermCountEnum.java:
```diff
@@ -0,0 +1,121 @@
+    @Override
+    public BytesRef next() throws IOException {
+        index++;
+        if(index >= sortedTerms.length) {
```
```suggestion
        if (index >= sortedTerms.length) {
```

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TermCount.java:
```diff
@@ -0,0 +1,107 @@
+    );
+    static {
+        PARSER.declareString(optionalConstructorArg(), new ParseField(TERM_FIELD));
+        PARSER.declareLong(optionalConstructorArg(), new ParseField(DOC_COUNT_FIELD));
```
Should we be using `constructorArg()` rather than `optionalConstructorArg()`?

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TermCount.java:
```diff
@@ -0,0 +1,107 @@
+
+    void setDocCount(long docCount) {
+        this.docCount = docCount;
+    }
```
I'm not seeing where these setters are called, are they needed? If not can we make term and docCount final?

markharwood replied:
addToDocCount(int) is used to combine counts from shards but the setDocCount isn't. Removed.

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TermEnumRequest.java:
```diff
@@ -0,0 +1,175 @@
+        builder.field("field", field);
+        builder.field("string", string);
+        builder.field("size", size);
+        builder.field("timeout", timeout().getMillis());
```
let's write it as a time value?

jpountz commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/termenum/MultiShardTermsEnumTests.java:
```diff
@@ -0,0 +1,126 @@
+                for (DirectoryReader reader : readers) {
+                    Terms terms = MultiTerms.getTerms(reader, fieldName);
+                    TermsEnum te = automaton.getTermsEnum(terms);
+                    if(randomBoolean()) {
```
```suggestion
                    if (randomBoolean()) {
```

jpountz commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/termenum/MultiShardTermsEnumTests.java:
```diff
@@ -0,0 +1,126 @@
+                        // Simulate fields like constant-keyword which use a SimpleTermCountEnum to present results
+                        // rather than the raw TermsEnum from Lucene.
+                        ArrayList<TermCount> termCounts = new ArrayList<>();
+                        while(te.next()!=null) {
```
```suggestion
                        while (te.next()!=null) {
```

jpountz commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/termenum/MultiShardTermsEnumTests.java:
```diff
@@ -0,0 +1,126 @@
+                HashMap<String, Integer> expecteds = new HashMap<>();
+
+                for (String term : globalTermCounts.keySet()) {
+                    if(term.startsWith(searchPrefix)) {
```
```suggestion
                    if (term.startsWith(searchPrefix)) {
```

jpountz commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/termenum/MultiShardTermsEnumTests.java:
```diff
@@ -0,0 +1,126 @@
+                MultiShardTermsEnum mte = new MultiShardTermsEnum(termsEnums.toArray(new TermsEnum[0]));
+                HashMap<String, Integer> expecteds = new HashMap<>();
+
+                for (String term : globalTermCounts.keySet()) {
```
nit: can you iterate over `entries` rathen than keys since you need values too?

jpountz commented on x-pack/plugin/mapper-constant-keyword/src/main/java/org/elasticsearch/xpack/constantkeyword/mapper/ConstantKeywordFieldMapper.java:
```diff
@@ -140,6 +144,20 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
+            if (matches == false) {
+                return null;
+            }
+            int docCount = queryShardContext.searcher().getIndexReader().numDocs();
```
should we use `maxDoc()` to be consistent with keyword fields, which don't ignore deletes?

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java:
```diff
@@ -0,0 +1,552 @@
+            }
+
+        } catch (Exception e) {
+            error = e.getMessage();
```
I usually don't like losing stack traces, which can be very useful for debugging purposes.

markharwood commented on docs/reference/search/term-enum.asciidoc:
```diff
@@ -0,0 +1,98 @@
+[[search-term-enum]]
+=== Term enum API
```
I lose count myself of how many ways there are to do autocomplete. Maybe we need more docs to be updated e.g. https://www.elastic.co/guide/en/elasticsearch/reference/7.x/search-as-you-type.html

markharwood replied:
Now have `_terms_enum` REST endpoint and `TermsEnumxxx` classnames

markharwood commented on docs/reference/search/term-enum.asciidoc:
```diff
@@ -0,0 +1,98 @@
+  "complete" : true
+}
+--------------------------------------------------
+// TESTRESPONSE[s/8/$body.terms.0.doc_count/]
```
copy/paste relic I suspect

markharwood commented on docs/reference/search/term-enum.asciidoc:
```diff
@@ -0,0 +1,98 @@
+// TESTRESPONSE[s/8/$body.terms.0.doc_count/]
+
+The "complete" flag is false if time or space constraints were met and the
+set of terms examined was not the full set of available values.
```
When this PR had a sort-by-popularity mode we had a space constraint of only considering the first 10k terms+counts with the prefix. Whatever modes and constraints we place in future the one thing end users need to know is "is it worth finishing typing the word in my head if it's not shown in the suggestions so far?". The reason(s) why we haven't yet shown them what's in their head don't matter too much.

jimczi replied:
That makes me think that we should also add an option to search after a value. That would allow to paginate values that match a prefix efficiently ?

markharwood replied:
Yep. A search_after would make sense

markharwood commented on docs/reference/search/term-enum.asciidoc:
```diff
@@ -0,0 +1,98 @@
+[[term-enum-string-param]]
+`string`::
+(Mandatory, string)
+The string to match at the start of indexed terms
```
We opted for `string` to allow a future "match_mode" style flag to change behaviour e.g. infix matching

markharwood committed:
Addressing some review comments (thanks Jim/Adrien!)

markharwood committed:
Docs tidy up

markharwood committed:
Provide full stack traces for errors, change TODO comment

markharwood commented:
This is currently failing license usage tests because running the new `10_basic.yml` test triggers some security checks which are unexpected by the existing [10_feature_usage.yml ](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/core/src/yamlRestTest/resources/rest-api-spec/test/license/10_feature_usage.yml) startup test.

Maybe there's something that needs to be done to the sequencing and isolation of these tests? Thoughts @rjernst ?

markharwood committed:
Move location of YAML test - was causing errors when seated alongside core/src/yamlRestTest

markharwood committed:
Security enhancement - allow access where DLS rewrites to match_all. Added security tests

markharwood committed:
Remove acquisition of searcher from security check code

markharwood committed:
Changed termenum to termsenum. REST endpoint is now _terms_enum

markharwood committed:
Checkstyle fix

markharwood commented on docs/reference/search/term-enum.asciidoc:
```diff
@@ -0,0 +1,98 @@
+(Optional, integer)
+The maximum length of time in milliseconds to spend collecting results. Defaults to 1000.
+If the timeout is exceeded the `complete` flag set to false in the response and the results may
+be partial or empty.
```
Will add a max limit e.g. 1 minute

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termsenum/action/TransportTermsEnumAction.java:
```diff
@@ -0,0 +1,599 @@
+        }
+        if (request.shardIds().size() == 0) {
+            listener.onResponse(new NodeTermsEnumResponse(request.nodeId(), Collections.emptyList(), null, true));
+        } else {
```
We should use the search threadpool here if the queue is empty.
Something like:
```java
        assert transportService.getThreadPool().executor(ThreadPool.Names.SEARCH) instanceof EsThreadPoolExecutor
            : "SEARCH threadpool must be an instance of ThreadPoolExecutor";
        EsThreadPoolExecutor ex = (EsThreadPoolExecutor) transportService.getThreadPool().executor(ThreadPool.Names.SEARCH);
        final String executorName = ex.getQueue().size() == 0 ? ThreadPool.Names.SEARCH : shardExecutor;
```

markharwood committed:
Addressing review comments - formatting, thread pool choices and more

markharwood committed:
Oops. Thought I’d resolved this review comment but hadn’t

markharwood committed:
Changed timeout setting to a TimeValue

markharwood committed:
Checkstyle fix

markharwood committed:
In flattened fields make only the value (not the field name) subject to case insensitive search

markharwood commented:
I think I've addressed all the review comments now except for your question about timers @jpountz , which I've commented on [here](https://github.com/elastic/elasticsearch/pull/66452#discussion_r621052986)

markharwood committed:
Moved initialisation of data node timing of request from NodeTermsEnumRequest constructor to TransportTermsEnumAction#asyncNodeOperation

markharwood commented:
I think that's all the review comments addressed now, @jpountz + @jimczi (thanks for all those!)

jimczi commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termsenum/action/TransportTermsEnumAction.java:
```diff
@@ -0,0 +1,610 @@
+            } else {
+                try {
+                    // TODO pass through a reduced timeout (the original time limit, minus whatever we may have
+                    // spent already getting to this point.
```
Is the TODO still needed ?

jimczi reviewed (approved):
LGTM!

markharwood committed:
Remove outdated TODOs
